controllers/configurationpolicy_utils.go

```diff
@@ -577,98 +577,161 @@ func createStatus(
+		names := msgMap[msgTemplate]
+		slices.Sort(names) // Sort names for consistent output
+
+		compliancyDetailsMsg = setCompliancyDetailsMsgEnd(compliancyDetailsMsg)
+
+		joinedNames := strings.Join(names, ", ")
+		if joinedNames != "" {
+			joinedNames = fmt.Sprintf(" [%s]", joinedNames)
+		}
+
+		compliancyDetailsMsg += fmt.Sprintf(msgTemplate, resourceName, joinedNames)
+	}
+
+	return compliancyDetailsMsg
+}
 
-			// This prevents repeating the same reason for each unique object name list.
+// Process the object name strings in order to ensure a deterministic reason and message.
+// This function returns msgMap, which is a map where the key is a message template
+// and the value is a list of resource names that share the same message template.
+func generateMessageWithReason(sortedObjectNamesStrs []string,
+	objectNameStrsToNsNames map[string][]string,
+	nsNameToEvent map[string]*objectTmplEvalResultWithEvent,
+	compliancyDetailsReason string,
+	compliancyDetailsMsg string,
+	reason string,
+) (map[string][]string, string, string) {
+	msgMap := map[string][]string{}
+
+	for i, namesStr := range sortedObjectNamesStrs {
+		if compliancyDetailsMsg != "" && !strings.HasSuffix(compliancyDetailsMsg, "; ") {
+			compliancyDetailsMsg += "; "
+		}
+
+		var generatedReason, msgTemplate string
+
+		// Generate reason and message template based on the 'reason' value.
+		switch reason {
+		case reasonWantFoundExists:
+			generatedReason = "K8s `must have` object already exists"
+			msgTemplate = "%s%s found as specified"
+		case reasonWantFoundCreated:
+			generatedReason = reasonWantFoundCreated
+			msgTemplate = "%s%s was created successfully"
+		case reasonUpdateSuccess:
+			generatedReason = reasonUpdateSuccess
+			msgTemplate = "%s%s was updated successfully"
+		case reasonDeleteSuccess:
+			generatedReason = reasonDeleteSuccess
+			msgTemplate = "%s%s was deleted successfully"
+		case reasonWantFoundDNE:
+			generatedReason = "K8s does not have a `must have` object"
+			msgTemplate = "%s%s not found"
+		case reasonWantFoundNoMatch:
+			generatedReason = "K8s does not have a `must have` object"
+			msgTemplate = "%s%s found but not as specified"
+		case reasonWantNotFoundExists:
+			generatedReason = "K8s has a `must not have` object"
+			msgTemplate = "%s%s found"
+		case reasonWantNotFoundDNE:
+			generatedReason = "K8s `must not have` object already missing"
+			msgTemplate = "%s%s missing as expected"
+		default:
+			// If it's not one of the above reasons, then skip consolidation. This is likely an error being
+			// reported.
 			if i == 0 {
 				if compliancyDetailsReason != "" {
 					compliancyDetailsReason += "; "
 				}
 
-				compliancyDetailsReason += generatedReason
+				compliancyDetailsReason += reason
 			}
 
-			compliancyDetailsMsg += generatedMsg
+			for j, nsName := range objectNameStrsToNsNames[namesStr] {
+				if j != 0 {
+					compliancyDetailsMsg = setCompliancyDetailsMsgEnd(compliancyDetailsMsg)
+				}
```

Comment:
If `j == 0`, then the current message would be empty, right? So then `setCompliancyDetailsMsgEnd` would not add a semicolon?

I think checking the index and the message in `setCompliancyDetailsMsgEnd` is redundant. I'd prefer to just use the index, and add the semicolon directly:

```suggestion
			for j, nsName := range objectNameStrsToNsNames[namesStr] {
				if j != 0 {
					compliancyDetailsMsg += "; "
				}
```